### PR TITLE
Jottacloud resume / de-duplication support

### PR DESCRIPTION
**Pull request by olihey**
Since we have to calculate the MD5 before uploading a file I added support for talking to the new API from Jottacloud to pre-allocate a file and get some information back before uploading.
The allocate method returns http status code 201 when the file is already uploaded and we can skip uploading it altogether (de-duplictation). A copy is still created at the desired location.
Additionally it also returns a "resume" position for the file if the upload was interrupted before. That way rclone can resume uploads to Jottacloud now with this PR.

@ncw I am having some trouble with the oauth for the api server from Jottacloud. They require the string for the `grant_type` to be uppercase. In the code https://github.com/ncw/rclone/blob/7194c358ad7b0287cdc44081a98879349d819288/vendor/golang.org/x/oauth2/oauth2.go#L244 for the oauth2 library it's lowercase. And using lowercase results in the API returning a 400 Bad Request.
Any idea how I could fix that? Do I need to write my very own ouath2.Client which would basically be a 1:1 copy of the existing one, just with the strings for `grant_type` changed to uppercase?

### CODE REVIEW

**Comment by olihey** [comment]
I fixed the grant_type issue by using a filter that catches refresh requests and then just changes the case.

One thing I am uncertain about how to handle. With this PR an oauth token is stored into the config for the remote. Existing remotes don't haven that setup. **Would it be okay to require users to setup the remote again?** I mean there shouldn't be to many users as support for Jottacloud didn't even made it's way into the "stable" release

**Comment by jkaberg** [comment]
@olihey Undoubtly the new API is the way going forward for Jotta. How about migrating the rest of this backend to the new API aswell? Just thinking out loud..

Perhaps @buengese has some thoughts?

**Comment by olihey** [comment]
Well sure. But the official clients from Jottacloud still use the mix of new API and "old" JFS and for rclone there isn't really anything to gain these days from the Jottacloud API. Browsing and downloading files is not possible at the moment with the API endpoint.
Biggest problem is the missing documentation and that there is no way to apply for an official API key. The one I am using in rclone right now is one I created myself using the API. If they disbale this service and blacklist the key we are screwed 😄 
For anyone interested, I got the info from here: https://github.com/ttyridal/aiojotta/wiki/Jotta-protocol-3.-Authentication

**Comment by olihey** [comment]
@buengese I agree with you on the danger of this. But overall even the JFS enpoints are not officially support and just documented in a forums post by one of the Jottacloud people. In the end we are screwed anyway if they change stuff. Just look at the MD5 before upload issue we has several days ago.

Maybe it should be noted in the documentation that this is UNOFFICIAL

**Comment by ncw** [comment]
@olihey wrote:
> I  fixed the grant_type issue by using a filter that catches refresh requests and then just changes the case.

Very innovative - I like it :-)

> One thing I am uncertain about how to handle. With this PR an oauth token is stored into the config for the remote. Existing remotes don't haven that setup. Would it be okay to require users to setup the remote again? I mean there shouldn't be to many users as support for Jottacloud didn't even made it's way into the "stable" release

Jottacloud has now made it into the stable release.  However it isn't unknown for remotes to need re-authenticating on a new release.  We had to do this with the onedrive backend and graph recently so as long as we print a clear message to the user I don't mind.

As for the undocumented API... Maybe using the new API could be optional? WIth a config parameter maybe if we aren't sure whether it is stable.

**Comment by olihey** [comment]
> As for the undocumented API... Maybe using the new API could be optional? With a config parameter maybe if we aren't sure whether it is stable.

My opinion on this subject is, **there is no official API documentation for ANY Jottacloud endpoint**. Everything we do in rclone when working with Jottacloud can break at any moment. No matter if it is the JFS endpoints or the new API endpoint.
Prime example is the required MD5 hash before upload.
I have posted a [question](http://forum.jotta.no/jotta/topics/using-the-api) about using the API in the official Jottacloud forum but nothing has happend yet.

Sure we could make the use of the new API optional but I don't really see the point since even the JFS endpoint can go away/break/change tomorrow without any notice.

**Comment by ncw** [comment]
@olihey sorry I have dropped the ball on this PR :-(

As far as I can see this is ready to merge with the one possible question mark of it needing a re-auth of the Jottacloud backend.

My question would be should we merge this for v1.44 (due 13 October)?

**Comment by ncw** [comment]
It needs rebasing to fix a conflict first though!

**Comment by olihey** [comment]
I'll have look as soon as I find time 😸

**Comment by olihey** [comment]
Sorry for the delay, should be fine now

**Comment by ncw** [comment]
I'm going to hold this for the v1.44 release then merge next week so it can get a bit more testing before a release.

**Comment by olihey** [comment]
Just got this feedback from the Jottacloud team:
> Hello,  we are currently investigating the road to offering some public API's. We will probably not offer the current API's as public, but we will definitely offer a set of API's for external developers so they can create a meaningful app integration with Jottacloud. 
​At this point I can't really give you a timeframe for this, but Jottacloud will work to adopt an open API strategy.

So, YEAH!!!!

**Comment by ncw** [comment]
> Just got this feedback from the Jottacloud team:
> 
> > Hello,  we are currently investigating the road to offering some public API's. We will probably not offer the current API's as public, but we will definitely offer a set of API's for external developers so they can create a meaningful app integration with Jottacloud.
> > ​At this point I can't really give you a timeframe for this, but Jottacloud will work to adopt an open API strategy.
> 
> So, YEAH!!!!

:-)

I realise I've let this PR drop off my radar - sorry!

I think it is ready to merge - is that right?

**Comment by olihey** [comment]
@ncw Should be ready. I am using this for weeks now without any issues :)

**Comment by ncw** [comment]
I've rebased this branch on master, squashed the commits and fixed all the merge conflicts and uploaded back to your PR branch.  (I suggest you pull it back to a new branch so you have the original branch.)

I think the code changes look fine.  The new auth requires a reconfig which is signalled properly by error messages which is great.

I then ran the backend integration tests

```
go install github.com/ncw/rclone/fstest/test_all
test_all -backends jottacloud
```

To get this.  So a few test failures [which are currently passing](https://pub.rclone.org/integration-tests/current/).  So I think this is a regression.  It is possible I messed up the rebase - the conflicts where a little tricky so if I've introduced the problem - my apologies!

image

Attached is the error log for the one of the failed tests.  (The other looks identical).

[jottacloud-fs.sync-TestJottacloud-5.txt](https://github.com/ncw/rclone/files/2568547/jottacloud-fs.sync-TestJottacloud-5.txt)

**Comment by olihey** [comment]
Hej, I probably won't have time for this in 2018. Sorry

**Comment by olihey** [comment]
@buengese If you have the time you are welcome to fix it. I was planning on looking at this around mid if January. No time before :(

**Comment by buengese** [comment]
All tests should pass now. It was just a minor mistake that took me way to long find. I also made some small changes to config. If we're using token based auth there is no need to store users password anymore. This will also allow us to implement 2fa which I'll do in a follow up PR.

**Comment by buengese** [comment]
@ncw Don't want to bother you but could you maybe have a quick look over the code so wen can merge this now. It's been ready for 2 weeks now.

File: backend/jottacloud/jottacloud.go [hunk 1]
```diff
@@ -261,6 +342,38 @@ func (o *Object) filePath() string {
 	return o.fs.filePath(o.remote)
 }
 
+// dummyClose is required for the grantTypeFilter below
```
[the remark this answers is not in the document]
Comment (ncw) [reply backend/jottacloud/jottacloud.go]: Not sure you've see this comment?

**Comment by buengese** [comment]
Thanks redoing auth is expected but there should have been a better error message. Looks like i accidentally broke this.
Jottacloud OAuth implementation is very much non standard in the sense that there is no OAuth page. I'm not sure if they plan this for future given that the switch to OAuth in the official Client is rather recent. For now we're replicating what the official client. The Password is only needed during setup and is no longer stored in the config (as opposed to the old implementation when we used basic user, password authentication)

**Comment by ncw** [comment]
> Thanks for the review!
> 
> > I had to redo the auth for my existing remotes - is this expected? Can we detect that somehow and tell the user rather than let them get this error message?
> 
> Redoing auth is expected but there should have been a better error message. Olihey implementation already did this correctly. Looks like i accidentally broke this somehow.

Ooops!  Are you going to fix this up in this PR?
 
> > Why do we ask the user for the password - can't we use a standard oauth flow?
> 
> Jottacloud OAuth implementation is very much non standard in the sense that there is no OAuth page. I'm not sure if they plan this for future given that the switch to OAuth in the official Client is rather recent. For now we're replicating what the official client does. The Password is only needed during setup and is no longer stored in the config (as opposed to the old implementation when we used basic user, password authentication)

That makes sense - thanks for explaining.

**Comment by ncw** [comment]
> Finally had some time again. I switched to `ioutil.NopCloser`

Great

> I cannot however reproduce your error message the one I get is
> 
> ```
> 2019/01/26 18:06:19 Failed to create file system for "jtest:": Failed to configure Jottacloud oauth client: empty token found - please run rclone config again
> ```
> which is the one 1 would expect.

I think what must have happened is I used my remote config for an intermediate version which has an incompatible token.

I'm going to merge this now - thank you very much :-)  So it will be in the 1.46 release.

**Pull request merged**